Deconstructing AIs

13spacemen opened the pull request:
Currently being turned into an AI is a 1-way process, you can't go back to being human while borgs can

This PR lets you deconstruct an AI and get the MMI out, this includes roundstart AIs so you can turn your doorknob into a real boy

- [x] Make sure all antag datums and memories and stuff are transferred
- [x] Make sure it doesn't break anything that relies on AIs
- [x] Make the AI combat upgrade drop, if installed
- [x] Checks for carded and dead and malf AIs
- [x] Add examine hints to AI core (de)construction

## Why?
Emergent gameplay, also it's unintuitive that you can't deconstruct AIs while you can borgs.

## State of the PR

- You can now deconstruct AIs, examine them for construction hints
- Use crowbar to open the access panel, then wirecutters on the AI, if the AI gives consent you don't need roboticist access, if they reject you'll need access or to emag the AI or to kill the AI
- Dead AIs take very long to deconstruct, (currently a flat 80 seconds)
- If AI had the combat upgrade, it drops on deconstruction
- Malf AIs will shock you into crit if you try to deconstruct them
- AI core (de)construction is more user friendly, you can examine it for hints and it uses balloon alerts
- You can now insert inactive brains into MMIs and inactive MMIs/posibrains into AI cores (it will just make an inactive core, to use an intellicard on, not an AI mob)
- AI construction must require an MMI/posibrain now, you can't insert reinforced glass without one (empty ones work, if making an empty core)
- AIs who get gibbed drop an MMI/posibrain, if it was malf or had the combat upgrade, it drops that

:cl:
add: You can now deconstruct living AIs and turn them back into AI core frames with MMIs/posibrains. The AI can finally be a real boy
qol: AIs that are gibbed now drop MMIs/posibrains
/:cl:

13spacemen committed:
adds ability to use wirecutters on AI mob to turn it into an AI core structure with an MMI in it

JohnFulpWillard commented:
1- I was certain AIs were able to be deconstructed already, roundstart ones are MMIs
2- Wouldn’t this turn Posibrain AIs into a regular brain?

13spacemen commented:
> 1- I was certain AIs were able to be deconstructed already, roundstart ones are MMIs 2- Wouldn’t this turn Posibrain AIs into a regular brain?

they can't be deconstructed, you can deconstruct the /obj/structure/ai_core but once it becomes /mob/living/silicon/ai there's no way back

as for posibrains, yeah need to sort out all those little issues

Shadowflame909 commented:
I feel like this kind of sidesteps the purpose of intellicards. Don't they have their own loop of taking forever to permanently wipe to kill an ai? Now you can decon one and space it's brain

13spacemen commented:
@Shadowflame909 can't you just space the AI by unwrenching it, or just space the intellicard?

Mothblocks commented:
can you please change the title of your pr it isnt obvious at all

13spacemen committed:
ai combat upgrade drops if installed
ais drop posibrains if built with them
dead AIs cant be deconstructed
borg shells disconnected, doomsday device shut off if deconstructed
ai_to_mmi proc

zxaber commented:
I disagree with this change. AIs are supposed to be a program on a machine, not a traditional entity (ignoring the handling of the AI mob itself). This is why they can fit on a card, can shunt themselves into APCs, etc.

It's a one-way process, the brain is destroyed to make a software map.

MMMiracles commented:
i kinda just assumed AIs were like wetware computers that used the brain as a foundation and the personality/AI itself was the software that ran on the brain, hence why it could be carded if needed.

i feel like being able to reverse the process should be possible but there should be a permanent downside to having done it, whether it be like requiring a donor brain to 'upload' the AI into via the MMI or permanent EMP vulnerability.

13spacemen committed:
examine hints and balloon alerts for ai core (de)construction
all built ai cores must have an mmi/posibrain, even empty ones
AI_READY_CORE ai cores can be activated with screwdriver if mmi has mind
disconnecting monitor needs wirecutters instead of screwdriver
fixed some misc runtimes
getting gibbed as AI drops an MMI/posibrain
mmis allow inserting inactive brains
ais have examine hints for those with roboticist access or if its emagged
malf ais will crit electrocute those trying to deconstruct
you can deconstruct AI without consent by having roboticist access or emagging it

Jacquerel commented:
> AIs should be less humanised not more.

I know it's just #lore but if we wanted to disassemble AIs I personally would prefer them to default to posibrains for roundstart ones and release brains only if they were constructed during the round from brains, rather than the other way around.

zxaber commented:
>all built ai cores must have an mmi/posibrain, even empty ones

Are you saying that it is now impossible to build an empty core without a spare brain? That's a bit dumb if you're in a situation where you have an AI carded and need to make a new core.

13spacemen commented on code/modules/antagonists/heretic/knowledge/starting_lore.dm:
```diff
@@ -87,6 +87,8 @@ GLOBAL_LIST_INIT(heretic_start_knowledge, initialize_starting_knowledge())
 		to_chat(user, span_boldnotice("You don't have a heart, or any chest organs for that matter. You didn't get a Living Heart because of it."))
 
 /datum/heretic_knowledge/living_heart/on_lose(mob/user)
+	if(!IS_HERETIC(user))
+		return
```
To test the antag datums stuff, I gave myself multiple antags via Traitor Panel and then clicked Transform to AI
most of it was flawless, traitor, rev, changeling, thief
this is one of the runtimes i got, it didn't actually cause any issues, i humanized the AI and defibbed the human body and i could still use my heretic powers, but to get rid of the runtime i put this check here

13spacemen commented on code/modules/antagonists/heretic/heretic_knowledge.dm:
```diff
@@ -184,6 +184,8 @@
 	user.mind.AddSpell(spell_to_add)
 
 /datum/heretic_knowledge/spell/on_lose(mob/user)
+	if(!user.mind)
+		return
```
this is another one of the runtimes i got, not quite sure why this happens, might be because during testing i stacked a lot of antags on top of each other, will try with only heretic though

13spacemen replied:
@MrMelbert there is an issue with heretics transferring bodies, its not an AI-exclusive thing, kept getting these 2 runtimes where it asks for the mind of the old body but that gets set to null in /datum/mind/transfer_to()

13spacemen replied:
removed all the unnecessary changes to files @MrMelbert

13spacemen committed:
prevent a space ruins runtime

Mothblocks commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -100,43 +128,57 @@
+/obj/structure/ai_core/screwdriver_act(mob/living/user, obj/item/tool)
+	. = ..()
+	if(state == AI_READY_CORE && brain?.brainmob?.mind)
+		balloon_alert(user, "connecting neural network")
```
```suggestion
		balloon_alert(user, "connecting neural network...")
```

Mothblocks commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -100,43 +128,57 @@
 				return
 
-			to_chat(user, span_notice("You start to deconstruct the frame..."))
+			balloon_alert(user, "deconstructing frame")
```
```suggestion
			balloon_alert(user, "deconstructing frame...")
```

Mothblocks commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -197,7 +245,16 @@
+							if(!user.transferItemToLoc(M, src))
+								return
+							brain = M
+							balloon_alert(user, "added [M.name] to frame")
```
```suggestion
							balloon_alert(user, "added mmi to frame")
```

We don't want to use custom names in balloon alerts to guarantee they are short.

Mothblocks commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -223,33 +280,17 @@
 				if(P.tool_behaviour == TOOL_CROWBAR)
 					P.play_tool_sound(src)
-					to_chat(user, span_notice("You remove the glass panel."))
+					balloon_alert(user, "removed glass panel.")
```
```suggestion
					balloon_alert(user, "removed glass panel")
```

Mothblocks commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -208,13 +265,13 @@
 
 					brain = M
-					to_chat(user, span_notice("You add [M.name] to the frame."))
+					balloon_alert(user, "added [M.name] to frame")
```
```suggestion
					balloon_alert(user, "added mmi to frame")
```

Mothblocks commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -259,14 +300,38 @@
+		return FALSE
+	B.mind?.remove_antags_for_borging()
+
+	var/mob/living/silicon/ai/A = null
```
Better variable name

Mothblocks commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -259,14 +300,38 @@
 	return ..()
 
+/obj/structure/ai_core/proc/ai_structure_to_mob()
+	var/mob/living/brain/B = brain.brainmob
```
Better variable name

Mothblocks commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -259,14 +300,38 @@
+	if(brain.braintype == "Android")
+		A.posibrain_core = TRUE
+	SSblackbox.record_feedback("amount", "ais_created", 1)
+	deadchat_broadcast(" has been brought online at <b>[get_area_name(A, TRUE)]</b>.", span_name("[A]"), follow_target=A, message_type=DEADCHAT_ANNOUNCEMENT)
```
Named parameter for the `TRUE` in `get_area_name`. Also this looks like a copy post.

13spacemen replied:
@Mothblocks This is the same code that would execute currently if you used a screwdriver on a `state == GLASS_CORE` AI, i just moved it to a separate proc

Mothblocks commented on code/modules/mob/living/silicon/ai/ai.dm:
```diff
@@ -377,6 +381,82 @@
+	if(hack_software)
+		new/obj/item/malf_upgrade(get_turf(src))
+	the_mmi = mmi_type
+	the_mmi.brain = new /obj/item/organ/internal/brain(the_mmi)
```
Does this transfer suicide?

Mothblocks reviewed (changes requested):
How does this handle a suicided AI having their brain taken out?

13spacemen committed:
adjusted some balloon alerts
text has checks for MMI/posibrain
renamed some vars
SSblackbox does not record screwdrivering an AI_READY_CORE AI

13spacemen committed:
moves var/emagged to /silicon level
named param for get_area_name()
changed ai deconstruction to be crowbar first then wirecutters
moved some procs to ai_defense.dm
added more checks
allow deconstructing dead non-malf AIs, but it takes 80 seconds
examine hints for ais improved, and no longer only for robotics access
set suiciding var on MMI generation for AIs and do not revive suicided brainmob

13spacemen committed:
adds more suiciding checks
cannot insert suicided brains into MMIs
suicided AIs suicide status transfers to the brain organ
set is_anchored var on death of AI

13spacemen committed:
undid unnecessary file changes

13spacemen commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -313,7 +388,7 @@ That prevents a few funky behaviors.
 
 /obj/structure/ai_core/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/item/aicard/card)
-	if(state != AI_READY_CORE || !..())
+	if(state != AI_READY_CORE || !..() || brain?.brainmob?.mind || brain?.brainmob?.suiciding)
```
@MrMelbert it is supposed to be `mind`, I changed AI core construction to always require a [non suicided] brain (even from monkeys)
We don't want to do "AI intellicard swipe on empty core" if that core's MMI has a mind (which it would if you reversed construction by deconstructing an AI)

13spacemen committed:
some minor changes
removed redundant vars and checks
tried to make code more readable
made ai (de)construction respect combat mode
renamed ai core brain var to core_mmi

13spacemen commented:
@MrMelbert did my best to make the code more readable some of the checks such as brainmob.mind and brainmob.suiciding together can't be avoided though since it's possible for a brainmob to have a mind and also be suicided
removed as many double ternary''s and redundancies as i could

13spacemen committed:
renames var/posibrain_core to posibrain_inside
removes useless var/can_deconstruct on ai cores that didn't prevent anything
carded AIs match the mmi/posibrain of whatever core they're uploaded to

13spacemen committed:
mind has new var/has_ever_been_ai
adjusted ssblackbox AIs created

Mothblocks commented:
> Using a surgery computer with an AI card slot to read the AI and "program" the brain of anotherwise braindead human(/humanized monkey) would be a neat reversal method

would be cool but i can't really require it on this pr
